euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/windowing/MergingWindowSet.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright 2016 Seznam.cz, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.seznam.euphoria.flink.streaming.windowing;
+
+import cz.seznam.euphoria.core.client.dataset.windowing.MergingWindowing;
+import cz.seznam.euphoria.core.client.dataset.windowing.Window;
+import cz.seznam.euphoria.core.client.util.Pair;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utility class to keep track of currently existing {@link Window} instances
+ * when using {@link MergingWindowing}.
+ *
+ * @param <W> type of the window
+ */
+class MergingWindowSet<W extends Window> {
+
+  private final MergingWindowing windowing;
```

Comment:
I don't think reusing this class would bring any benefit. It's too complicated because the window set must be persisted after each step in streaming. Most of the code is performance optimization to avoid costly persistent state allocation. This is not the case in `GroupReducer` where everything is processed in a plain `HashMap` in memory.